src/locales.cc

```diff
@@ -32,303 +32,214 @@
+      Fallback     = 1<<2, ///< Fallback of a requested locale (may also be requested)
+    };
 
-    tr << (*it).locale.code();
-    tr << (*it).locale.name();
-    tr << isRequestedState(it->state);
-    tbl << tr;
-  }
+    LocaleState()
+    {}
 
-  tbl.sort(1);
+    LocaleState( State state_r )
+    : _state { state_r }
+    {}
 
-  cout << tbl;
-}
+    bool isNone() const		{ return !_state; }
+    bool isRequested() const	{ return _state & Requested; }
+    bool isFallback() const	{ return _state & Fallback; }
 
-#if 0
-static void printXmlLocaleList( Zypper & zypper, const zypp::LocaleSet &locales )
-{
+    void tagRequested()		{ if ( !isRequested() ) _state |= Requested; }
+    void tagFallback()		{ if ( !isFallback() ) _state |= Fallback; }
 
-}
-#endif
-
-static void printLocalePackages( Zypper & zypper, const zypp::sat::LocaleSupport & myLocale )
-{
-  Table tbl;
-  TableHeader th;
+    std::string asString() const
+    { return isRequested() ? _("Requested") : isFallback() ? _("Fallback") : _("No"); }
 
-  // translators: header of table column - S is for 'Status' (package installed or not)
-  th << _("S");
-  // translators: header of table column - the name of the package
-  th << _("Name");
-  // translators: header of table column - the package summary
-  th << _("Description");
+  private:
+    unsigned _state = 0;
+  };
 
-  tbl << th;
+  using LocaleStateMap = std::map<Locale,LocaleState>;
 
-  for_( it, myLocale.selectableBegin(), myLocale.selectableEnd() )
+  /** 	lloc helper building the set of locales to show */
+  inline LocaleStateMap argsToLocaleSet( const std::vector<std::string> & localeArgs_r, bool showAll_r )
   {
-    TableRow tr(3);
-    zypp::ui::Status status = (*it)->status();
+    LocaleStateMap ret;
+    bool mustFixState = true;	// If locales a added in No state and later checked if Requested/Fallback
 
-    if ( status == zypp::ui::S_KeepInstalled || status == zypp::ui::S_Protected )
-      tr << "i";
-    else
-      tr << " ";
-    tr << (*it)->name();
-    tr << (*it)->theObj()->summary();
-
-    tbl << tr;
-  }
-
-  tbl.sort(1);
-
-  cout << tbl;
-}
-
-static zypp::LocaleSet relevantLocales( const std::vector<std::string> &localeArgs, bool relaxed )
-{
-  const zypp::LocaleSet & availableLocales( God->pool().getAvailableLocales() );
-
-  zypp::LocaleSet resultSet;
-
-  for_( it, availableLocales.begin(), availableLocales.end() )
-  {
-    if ( localeArgs.empty() )
+    if ( showAll_r )
     {
-      // without argument take only requested locales
-      if ( God->pool().isRequestedLocale(*it) )
+      for ( Locale rloc : God->target()->requestedLocales() )
+	ret[rloc].tagRequested();
+      for ( const auto & avloc : God->pool().getAvailableLocales() )
+	ret[avloc];	// state will be adjusted later
+    }
+    else if ( ! localeArgs_r.empty() )
+    {
+      // Search patterns are matched against available locales.
+      // Exact locales are added even if not in available locales.
```

Comment:
OTOH it's confusing if al locale I explicitely pass on the command line sometimes appears in the result and sometimes not. I asked for it so I want to see it and the state says 'Requested: No'.  Maybe we should differ between 'No' and 'n/a' in the output?

Reply:
```
$ zypper lloc en_XX en_US
Code  | Language                        | Requested
------+---------------------------------+----------
en_US | English (United States)         | No       
en_XX | English (Unknown country: 'XX') | No (n/a) 
```
Now locales being forced into the output just because they are passed on the CLI are tagged as not available.